designs/s3/design-vscode-s3.md

```diff
@@ -0,0 +1,262 @@
+          option. This is not S3-specific, it is used for all `Download`
+          experiences in the Toolkit.
+      - (P2) Loading bar or updating percentage.
+    - (P1) Toolkit tracks current downloads so that they are cancellable.
+      - User can cancel all downloads by clicking a button in the Toolkit
+        explorer toolbar, or by the `AWS: Cancel downloads` VSCode command.
+    - (P2) User can download multiple files/folders by selecting them in the S3
+      tree and using the `Download` context-menu item.
+  - (P0) `Download as...`
+    - Presents the standard file-chooser dialog. Otherwise behaves like the `Download` action.
+  - (P2) `Edit`
+    - Downloads the object to a temp folder and opens it in a [readonly view](https://code.visualstudio.com/api/extension-guides/virtual-documents).
+  - (P0) Upload... (similar UX as [S3 buckets](#s3-buckets) `Upload...` item)
+  - _Destructive operations (separator):_
+    - (P1) `Delete`
+      - If versioning is enabled, do _not_ prompt to confirm deletion.
+        - Show a passive toaster message: _Deleted `foo.txt`. `Undo` from previous versions?_
+        - If user clicks `Undo` in the toaster message, restore the file from the previous versions list.
+      - If versioning is not enabled, show a prompt to confirm deletion :(
+        - Show a passive toaster message: _Deleted `foo.txt`. (Cannot undo because S3 bucket versioning is disabled.)_
+
+### S3 bucket properties
+
+Surface these important bucket properties in a conspicuous yet uncluttered manner:
+
+- (P1) _Public accessible_: show icon (same as AWS web console) next to bucket
+  name indicating that it is publicly accessible.
+- (P1) _Versioning enabled_: show icon next to bucket name indicating that
+  versioning is enabled.
+
+### S3 object properties
+
+S3 object properties _size_, _last modified date_, and _previous versions_ (if
+versioning is enabled) are surfaced as follows:
+
+- (P0) _Size_ and _last modified date_ are exposed in a tooltip.
+  - Example: `42 kb, 2020-02-10 00:01:02 EST`
+- (P1) _Versions_: if versioning is enabled on the bucket and an object has previous versions:
+  - Mouse-hover on an item reveals an [inline](https://code.visualstudio.com/api/extension-guides/tree-view#view-actions) `Versions...` button
+    <br><img src="design-s3-hover-versions.png" width="200"/>
+  - Context-menu enables `Previous versions...` menu item
+    - (BLOCKED) grey-out or "disable" the menu item if bucket versioning is disabled or the item does not have previous versions available.
+  - Show an icon (count? ellipsis?) next to the object name, to indicate
+    that previous versions are available.  Example: `foo.txt (…)`
+  - Present versions as "children" of the object, but only when versions are explicitly requested by one of the above actions.
+
+### ([BLOCKED](https://github.com/microsoft/vscode/issues/32592)) Drag-drop from client to server
```

Comment:
I don't like these BLOCKED sections because there's zero indication that Microsoft will ever implement this feature. Assign a priority to these and we can re-evaluate if we want to add these in the future through a different mechanism if MS never adds it.

Reply:
Priority is zero until they're unblocked.